### ISSUE TITLE
Typescript import/named with types in `d.ts`

### ISSUE THREAD

**Issue by ThomasdenH** [benmosher/eslint-plugin-import#1341]
I think I have found a bug with `import/named`. I have created a [repository](https://github.com/ThomasdenH/eslint-plugin-import-named-bug) that reproduces the issue.

The line that gives problems is

```typescript
import { AreaProps } from 'recharts';
```

It imports an interface defined in a `d.ts` file. The following warning is given:

```
1:10  error  AreaProps not found in 'recharts'  import/named
```

PR #1304 should solve importing types to typescript, but maybe this doesn't work for declaration files? Or is there something wrong with my configuration?

**Comment by manuelbieh** [comment]
Struggling with the same issue and `express-serve-static-core` from `node_modules/@types/express-serve-static-core`. 

```js
import { RequestHandler, Request } from 'express-serve-static-core';
```
results in an error:
```sh
8:41  error  Unable to resolve path to module 'express-serve-static-core'  import/no-unresolved
```

**Comment by pcorpet** [comment]
The problem seems to come for types installed through http://definitelytyped.org/.

**Comment by benmosher** [comment]
can you try using https://www.npmjs.com/package/eslint-import-resolver-typescript as your resolver and see if it fixes?

**Comment by benmosher** [comment]
@manuelbieh I think your case is a little different than OP. I see you're using the typescript resolver but it works fine for me... and `no-unresolved` should report based on the existence of the true code file and not the types.

it would be possible for the `typescript` resolver to try a `@types/...` resolution but that would be in scope for that resolver, and not this plugin.

**Comment by pcorpet** [comment]
@benmosher , thanks the typescript resolver worked for me. Note for reference that you can have multiple import resolvers (I didn't want to turn off the webpack resolver so I hadn't tried the typescript one earlier).

**Comment by ThomasdenH** [comment]
> 
> 
> @ThomasdenH can you try using https://www.npmjs.com/package/eslint-import-resolver-typescript as your resolver and see if it fixes?
> 
> if it does, I think it may solve many TS-related issues. I just used it to solve some weird ones in my day job codebase
> 
> _edit:_ I think it won't help after all, based on [alexgorbatchev/eslint-import-resolver-typescript#17](https://github.com/alexgorbatchev/eslint-import-resolver-typescript/issues/17) + associated PR but it's an easy experiment

Sorry for the late reply!, Unfortunately it doesn't fix the issue, but thanks for the suggestion.

**Comment by benmosher** [comment]
@ThomasdenH: yep, makes sense. I didn't fully understand the scenario when I originally suggested.
@pcorpet: great! I'm doing the same thing -- I have the `webpack` resolver first for most things but the `typescript` resolver does resolution in some of my untranspiled first-party TS dependencies

**Comment by joelle-helgert-wittgruppe** [comment]
I am having the same issue for an intl-type that is generated by DefinitelyTyped. Is there any solution or further ideas on how to fix it?

**Comment by JounQin** [comment]
For anything maybe interested, you can try `eslint-import-resolver-typescript@JounQin/node_modules#eslint-import-resolver-typescript` temporarily (PR made at https://github.com/alexgorbatchev/eslint-import-resolver-typescript/pull/20).

**Comment by benmosher** [comment]
☝️ @JounQin's solution is my recommendation as the permafix. I haven't tried it myself but I scanned through the changes and they look like they should do the trick!

@JounQin: I'm not sure how to read that link. Did you publish your fork to `npm`?

**Comment by JounQin** [comment]
@benmosher It's just using npm package with github branch.

**Comment by dontsave** [comment]
I'm running into this issue as well. Imports from declarations are very common, so this is a real pain. It would be great to get a fix

**Comment by benmosher** [comment]
again, resolution is out of scope for this plugin. will close when @JounQin's PR is accepted or other similar fix is implemented by `eslint-import-resolver-typescript` (see https://github.com/benmosher/eslint-plugin-import/issues/1341#issuecomment-507016603 above)

**Comment by hedgepigdaniel** [comment]
The way I understand it, its not possible for a resolver to solve this problem. Resolvers can only return a single file, but in this case of separate source and declaration files, the exports are distributed across 2 files.

e.g.
```typescript
// file.d.ts
export type Fruit = {
  color: string,
};
```
```javascript
// file.js
export const tomato = {
  color: 'red',
};
```
```typescript
// index.ts
import { Fruit, tomato } from './file'; // <-- these exports do not come out of any one file
```

**Comment by JounQin** [comment]
@hedgepigdaniel You should add `export const tomato: Fruit` into `.d.ts` also.

**Comment by JounQin** [comment]
@benmosher Should this issue be closed in favor of [eslint-import-resolver-ts](https://github.com/rx-ts/eslint-import-resolver-ts)?

**Comment by w4-alibek** [comment]
Make sure you have inside the package.json
`
{
  ....
  "dependencies": {
    "module-name": "version"
  },
  "devDependencies": {
    ...
    "@types/module-name": "version" // version should be same with dependencies
  }
  ....
}
`

**Comment by JounQin** [comment]
@benfletcher @ThomasdenH @ljharb @stekycz 

This issue can be closed definitely in favor of [eslint-import-resolver-typescript@2.0.0](https://github.com/alexgorbatchev/eslint-import-resolver-typescript/tree/v2.0.0).

**Comment by hedgepigdaniel** [comment]
@JounQin which change in that new version addresses this issue?

**Comment by JounQin** [comment]
@hedgepigdaniel You can always read the CHANGELOG by yourself.

**Comment by ljharb** [comment]
@JounQin i've done so, and i have no idea which change addresses this issue. Could you clarify?

**Comment by JounQin** [comment]
@ljharb I was thinking we were in the same context here: https://github.com/alexgorbatchev/eslint-import-resolver-typescript/pull/20#issuecomment-530032941, didn't we?

**Comment by ljharb** [comment]
Yes but those don't tell me which specific commits in the changelog address this issue :-)

**Comment by JounQin** [comment]
@ljharb Did you read https://github.com/alexgorbatchev/eslint-import-resolver-typescript/blob/master/CHANGELOG.md#features?

> * resolve .ts/.tsx/.d.ts first, and then fallback to @types/* ([b11ede3](https://github.com/alexgorbatchev/eslint-import-resolver-typescript/commit/b11ede3c9fafbc548db011729fd64e958cde6e51))

**Comment by hedgepigdaniel** [comment]
Does that mean there is now a problem where if there is a declaration file, there is a false error when importing a value from the module that has the declarations?

**Comment by hedgepigdaniel** [comment]
I guess that isn't a problem if you have types for all exported symbols

**Comment by JounQin** [comment]
@hedgepigdaniel Sorry my English is not so good enough for understanding what's your point? Do you have an example?

**Comment by hedgepigdaniel** [comment]
```javascript
// apple.js

export const color = 'red';
export const shape = 'round';
```
```typescript
// apple.d.ts

export const color: string;
```
```typescript
// index.ts

// vv Error: shape is not defined in `apple.d.ts`!
import { shape } from './apple';
```

**Comment by JounQin** [comment]
@hedgepigdaniel Yeah, but that is a bad practice and should never happen.

**Comment by ljharb** [comment]
@JounQin it's not about what's your responsibility; it's about being friendly and helpful :-) Thanks for pointing me to https://github.com/alexgorbatchev/eslint-import-resolver-typescript/commit/b11ede3c9fafbc548db011729fd64e958cde6e51.

**Comment by JounQin** [comment]
> @JounQin it's not about what's your responsibility; it's about being friendly and helpful :-) Thanks for pointing me to [alexgorbatchev/eslint-import-resolver-typescript@b11ede3](https://github.com/alexgorbatchev/eslint-import-resolver-typescript/commit/b11ede3c9fafbc548db011729fd64e958cde6e51).

Emm... I'm always friendly, but my pointed CHANGELOG and PR both target the same commit which fixed this issue. Maybe we both need to be more patient and find out the key point  by ourselves? Cheers.
